File: dart/meetup.dart
class Meetup {
  static const mp = <String, int>{
    'Monday': 1,
    'Tuesday': 2,
    'Wednesday': 3,
    'Thursday': 4,
    'Friday': 5,
    'Saturday': 6,
    'Sunday': 7,
  };

  String meetup(
      {required int year,
      required int month,
      required String week,
      required String dayofweek}) {
    var dt = DateTime.utc(year, month, 1);
    final arr = <DateTime>[];
    while (dt.month == month) {
      if (dt.weekday == mp[dayofweek]) {
        arr.add(dt);
      }
      dt = dt.add(Duration(days: 1));
    }

    final indices = ['first', 'second', 'third', 'fourth'];

    if (indices.contains(week)) {
      return format(arr[indices.indexOf(week)]);
    }

    if (week == "last") {
      return format(arr.last);
    }

    for (final dt in arr) {
      if (dt.day >= 13) {
        return format(dt);
      }
    }

    throw 'unreachable';
  }

  String format(DateTime dt) {
    return "${dt.year}-${two_digits(dt.month)}-${two_digits(dt.day)}";
  }

  String two_digits(int n) {
    return n.toString().padLeft(2, '0');
  }
}

File: ocaml/meetup.ml
open Base
open CalendarLib

type schedule = First | Second | Third | Fourth | Teenth | Last

type weekday =
  | Monday
  | Tuesday
  | Wednesday
  | Thursday
  | Friday
  | Saturday
  | Sunday

(* result type (year, month, day_of_month) *)
type date = int * int * int

let equal = function
  | Date.Mon, Monday
  | Tue, Tuesday
  | Wed, Wednesday
  | Thu, Thursday
  | Fri, Friday
  | Sat, Saturday
  | Sun, Sunday ->
      true
  | _ -> false

let meetup_day schedule weekday ~year ~month =
  let start = ref @@ Date.make year month 1
  and finish = Date.make year (month + 1) 1 in
  let dates = ref [] in
  while Date.(!start < finish) do
    if equal (Date.day_of_week !start, weekday) then
      dates := !dates @ [ !start ];
    start := Date.add !start (Date.Period.day 1)
  done;
  let ans =
    match schedule with
    | First -> List.hd_exn !dates
    | Second -> List.nth_exn !dates 1
    | Third -> List.nth_exn !dates 2
    | Fourth -> List.nth_exn !dates 3
    | Last -> List.last_exn !dates
    | _ -> List.find_exn !dates ~f:(fun d -> Date.day_of_month d >= 13)
  in
  Date.(year ans, int_of_month @@ month ans, day_of_month ans)
